### PR TITLE
Improve Github Actions

### PR DESCRIPTION
**Pull request by avkos**
## Description

- All tests depend on the build. To decrease build time I've moved ESM and Types builds to a separate process
- Fix lint test to auto create .secret file if it doesn't not exists
- Replace the 14 to 18 Node version.

Please include a summary of the changes and be sure to follow our [Contribution Guidelines](../CONTRIBUTIONS.md).

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [ ] I have selected the correct base branch.
- [ ] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [ ] My changes generate no new warnings.
- [ ] Any dependent changes have been merged and published in downstream modules.
- [ ] I ran `npm run lint` with success and extended the tests and types if necessary.
- [ ] I ran `npm run test:unit` with success.
- [ ] I ran `npm run test:coverage` and my test cases cover all the lines and branches of the added code.
- [ ] I ran `npm run build` and tested `dist/web3.min.js` in a browser.
- [ ] I have tested my code on the live network.
- [ ] I have checked the Deploy Preview and it looks correct.
- [ ] I have updated the `CHANGELOG.md` file in the root folder.
- [ ] I have linked Issue(s) with this PR in "Linked Issues" menu.

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -23,24 +23,43 @@ jobs:
...
+      strategy:
+          matrix:
+              node: [ 18 ]
+      steps:
+          -   uses: actions/download-artifact@v3
+              with:
+                  name: web3-${{ matrix.node }}.js.tar.gz
+                  path: /tmp
+          -   run: tar -xf /tmp/web3-${{ matrix.node }}.js.tar.gz -C ./
+          -   run: test -f ./.secrets.json || echo '{"SEPOLIA":{"HTTP":"","WS":"","ACCOUNT":{"address":"0x0000000000000000000000000000000000000000","privateKey":""},"ALLOWED_SEND_TRANSACTION":false},"MAINNET":{"HTTP":"","WS":"","ACCOUNT":{"address":"0x0000000000000000000000000000000000000000","privateKey":""},"ALLOWED_SEND_TRANSACTION":false}}' > ./.secrets.json
```
Comment (Spacesai1or) [inline .github/workflows/build.yml]: The `.secrets` file is created after building the project using the [prebuild NPM script](https://github.com/web3/web3.js/blob/e550a9c88edec819a8ac0181fc2d321f8e959121/package.json#L36) which calls the [init.ts script](https://github.com/web3/web3.js/blob/4.x/scripts/init.ts)

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -23,24 +23,43 @@ jobs:
...
+      - run: yarn install --ignore-scripts
+      - run: yarn build:cjs
       - run: tar -czf /tmp/web3-${{ matrix.node }}.js.tar.gz --exclude="./.git" ./
       - uses: actions/upload-artifact@v3
         with:
           name: web3-${{ matrix.node }}.js.tar.gz
           path: /tmp/web3-${{ matrix.node }}.js.tar.gz
-
+  build-cjs-types:
+      name: Build ESM & Types
```
Comment (jdevcs) [inline .github/workflows/build.yml:35]: There was a ESM tests PR by @luu-alex , are these running after this build?

Comment (jdevcs) [reply .github/workflows/build.yml:35]: doesnt seem

File: .github/workflows/build.yml (hunk 1 above)
Comment (avkos) [reply .github/workflows/build.yml]: Thanks! I will change it properly

Comment (avkos) [reply .github/workflows/build.yml]: fixed

File: .github/workflows/build.yml (hunk 2 above)
Comment (avkos) [reply .github/workflows/build.yml:35]: change node version in esm tests.

**Comment by avkos** [comment]
Blackbox tests fixed
image
Here is a link to the tests: https://github.com/web3/web3.js/actions/runs/5018104015/jobs/8997618185

File: packages/web3-eth-accounts/src/account.ts [hunk 3]
```diff
@@ -67,6 +68,8 @@ import type {
 	SignResult,
 } from './types';
 
+const secp256k1 = ethereumCryptography.secp256k1 ?? ethereumCryptography;
+
```
Comment (Spacesai1or) [inline packages/web3-eth-accounts/src/account.ts]: What was the reason for needing this?

**Review by luu-alex — APPROVED** [review]

File: .github/workflows/build.yml [hunk 4]
```diff
@@ -14,7 +14,7 @@ jobs:
   build:
     strategy:
       matrix:
-        node: [14, 16]
+        node: [18, 16]
     name: Build Packages
```
Comment (jdevcs) [inline .github/workflows/build.yml]:
```suggestion
    name: Build CJS
```
lets be consistent with other build naming and more visibility in ci ?

File: .github/workflows/build.yml [hunk 5]
```diff
@@ -23,35 +23,66 @@ jobs:
...
+      - run: yarn install --ignore-scripts
+      - run: yarn prebuild
+      - run: yarn build:cjs
       - run: tar -czf /tmp/web3-${{ matrix.node }}.js.tar.gz --exclude="./.git" ./
       - uses: actions/upload-artifact@v3
         with:
           name: web3-${{ matrix.node }}.js.tar.gz
           path: /tmp/web3-${{ matrix.node }}.js.tar.gz
-
+  build-cjs-types:
```
Comment (jdevcs) [inline .github/workflows/build.yml]:
```suggestion
  build-esm-types:
```

File: .github/workflows/build.yml [hunk 6]
```diff
@@ -23,35 +23,66 @@ jobs:
...
+      name: Build ESM & Types
+      needs: build
```
Comment (jdevcs) [inline .github/workflows/build.yml:36]: there is no need of `needs: build`  so it should be removed? as ESM is not dependent on CJS technical

File: .github/workflows/build.yml [hunk 7]
```diff
@@ -23,35 +23,66 @@ jobs:
...
-          name: web3-14.js.tar.gz
-          path: /tmp
-      - run: tar -xf /tmp/web3-14.js.tar.gz -C ./
-      - run: yarn lint
+      name: Lint
+      needs: build
+      runs-on: ubuntu-latest
+      strategy:
+          matrix:
+              node: [ 18 ]
```
Comment (jdevcs) [inline .github/workflows/build.yml]: matrix should only be used when we want to run job for multiple variables, so it should be removed, as Lint is being run for one node version.

File: .github/workflows/build.yml [hunk 8]
```diff
@@ -23,35 +23,66 @@ jobs:
...
+          -   run: tar -xf /tmp/web3-${{ matrix.node }}.js.tar.gz -C ./
+          -   run: yarn lint
 
   build-web:
     name: Build Web
     needs: build
     runs-on: ubuntu-latest
+    strategy:
+        matrix:
+            node: [ 18 ]
```
Comment (jdevcs) [inline .github/workflows/build.yml]: same here, for matrix

File: .github/workflows/e2e_network_tests.yml [hunk 9]
```diff
@@ -5,18 +5,19 @@ on:
         - release/**
       tags:
         - v4.*
+
 jobs:
   build:
     strategy:
       matrix:
-        node: [14]
+        node: [16]
```
Comment (jdevcs) [inline .github/workflows/e2e_network_tests.yml]: same here for matrix

File: packages/web3-eth-accounts/src/account.ts (hunk 3 above)
Comment (jdevcs) [reply packages/web3-eth-accounts/src/account.ts]: in case of ESM `secp256k1` was undefined

**Comment by jdevcs** [comment]
Thanks @avkos for fixing ESM bug, and testing on v18, 
I suggested few improvements ( less priority ) as compared to other bugs like ( truffle migration bugs, tx, ..etc ) so we can do it later after full release if its time consuming.

File: .github/workflows/e2e_network_tests.yml (hunk 9 above)
Comment (avkos) [reply .github/workflows/e2e_network_tests.yml]: done

File: .github/workflows/build.yml (hunk 4 above)
Comment (avkos) [reply .github/workflows/build.yml]: done

File: .github/workflows/build.yml (hunk 5 above)
Comment (avkos) [reply .github/workflows/build.yml]: done

File: .github/workflows/build.yml (hunk 6 above)
Comment (avkos) [reply .github/workflows/build.yml:36]: esm build doesn't work without cjs

https://github.com/web3/web3.js/actions/runs/5027592934/jobs/9017264915?pr=6100

File: .github/workflows/build.yml (hunk 7 above)
Comment (avkos) [reply .github/workflows/build.yml]: done

File: .github/workflows/build.yml (hunk 8 above)
Comment (avkos) [reply .github/workflows/build.yml]: done

File: .github/workflows/build.yml (hunk 6 above)
Comment (Spacesai1or) [reply .github/workflows/build.yml:36]: @avkos It should though, our code is written using modules, so the CJS build should be irrelevant. I think when can keep this as is, but should create an issue to investigate further @jdevcs?

**Review by Spacesai1or — APPROVED** [review]
Assuming we create an issue to track above esm build error, LGTM

File: .github/workflows/build.yml (hunk 6 above)
Comment (avkos) [reply .github/workflows/build.yml:36]: https://github.com/web3/web3.js/issues/6108

**Pull request merged**